### PR TITLE
[Flow] Changes to dispatch region formation based on SDXL branch.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This PR merges changes to dispatch region formation that helped on reducing number of dispatches in the SDXL branch. It includes

1) Fuse `tensor.insert_slice` consumers in some cases. 2) Avoid fusing consumers that are dequantization operations (also
   tighten the definition of dequantization operations)
3) Add a mode to do aggressive fusion that assumes all operations
   within a dispatch are vectorized, avoiding some pessimizing fusion
   decisions which are made to avoid high stack allocations.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: What happens to this part? I don't see any tests about it.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: I can add a test.. basically such operations are cloned, so we dont want math operations cloned.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: Done!

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: Do we know which backend does not support the case? I think such vectorization should be easy to support today. We might not generate performant vectorized code, but we should be able to vectorize the conv op?

Is there an issue? Is anyone working on it?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: I see, it makes sense to me.. Would it be better if we only accept arith ops here? I.e., we can check if all the operations (without terminator) are arith ops or not.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:515]: nit: Check dilations here too

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: What about other kinds of conv ops? You could use `linalg::inferConvolutionDims` to get the strides and dilations for anything "convolution-like"

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:599]: What goes wrong for `tensor.empty`? That is the other common case I would expect.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: Maybe, but we need to see how this evolves. That might be what we need to do.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: I think for most backends we can vectorize this... I think if we can make this work on RoCM then we can use that to drive this forward....

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: I don't quite follow why we don't want to clone math ops. Is it because they are expensive? Then we might need a better heuristic for things like `math.absi`/`math.ceil`?

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:515]: I am walking this slowly a bit. This is under the "aggressive-fusion" flag that we start using with SDXL and more things here as we go. Eventually this method should just return true always. So No point generalizing this.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: I think we should hold the bar for fusion. We don't disable fusion just because one of the backends does not support it. The PR is probably fine because the number of dispatches is improved (though it may comes from other changes). Can you file an issue, assign it to someone, and add a TODO with the issue number?

**Review by hanhanW — APPROVED** [review]
LGTM if the TODO is added

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: this is kind of a black list that we grow if needed. Eventually this should just return true always.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:599]: inserting into an empty is kind of strange. But we can generalize as needed (as the comment says)

Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:599]: it happens very often for concatenations (same with insert_slice -> insert_slice). Anyway doesn't have to change in this PR.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:589]: Yes. I dont know what all will appear in what is known as a dequantization operation. I dont want to let things through in a vaccum. I think it would be better to evolve this on a need-basis.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:513]: The fusion changes have a chicken-and-egg effect. You need the backends to handle it before we can turn it on, and backend fixes dont happen without it being stressed due to fusion. So the best way forward is to land the fusion heuristic changes under a flag and get the backend to fix it. For now filed https://github.com/iree-org/iree/issues/17167

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:599]: yeah, maybe... I tend to slow walk such decisions... if you find a use case for `empty`, I can add it (will try anyway after I land this PR).

**Pull request merged**
